torch/fx/passes/split_module.py

```diff
@@ -162,27 +165,106 @@ def construct_graph(
         return base_mod_env, base_mod_attrs
 
     partitions: Dict[str, Partition] = {}
-    orig_nodes: Dict[str, torch.fx.node.Node] = {}
+    orig_nodes: Dict[str, Node] = {}
 
     def record_cross_partition_use(
-        def_node: torch.fx.node.Node, use_node: Optional[torch.fx.node.Node]
+        def_node: Node, use_node: Optional[Node]
     ):  # noqa: B950
-        def_partition_name = getattr(def_node, "_fx_partition", None)
-        use_partition_name = getattr(use_node, "_fx_partition", None)
-        if def_partition_name != use_partition_name:
-            if def_partition_name is not None:
-                def_partition = partitions[def_partition_name]
+        defined = getattr(def_node, "_fx_partition", None)
+        used = getattr(use_node, "_fx_partition", None)
+        if defined != used:
+            if defined is not None:
+                def_partition = partitions[defined]
                 def_partition.outputs.setdefault(def_node.name)
-                if use_partition_name is not None:
-                    def_partition.partition_dependents.setdefault(use_partition_name)
+                if used is not None:
+                    def_partition.dependents.setdefault(used)
 
-            if use_partition_name is not None:
-                use_partition = partitions[use_partition_name]
+            if used is not None:
+                use_partition = partitions[used]
                 use_partition.inputs.setdefault(def_node.name)
-                if def_partition_name is not None:
-                    use_partition.partitions_dependent_on.setdefault(def_partition_name)
+                if defined is not None:
+                    use_partition.dependencies.setdefault(defined)
+
+    def instantiate_node_partition_mapping(node):
+        partition_name = str(split_callback(node))
+
+        # add node to partitions
+        partition = partitions.get(partition_name)
+        if partition is None:
+            partitions[partition_name] = partition = Partition(partition_name)
+
+        partition.node_names.append(node.name)
+        node._fx_partition = partition_name
+
+    # Super Nodes are nodes which by their global state effects,
+    # "taint" all downstream nodes while they are active.
+    SUPER_NODE_TARGETS = [
+        torch.amp._enter_autocast,
+        torch.amp._exit_autocast,
+        torch._C._set_grad_enabled
+    ]
+
+    # For grad regions:
+    # ------------------------
+    # 1. first region: we do nothing
+    # 2. subsequent regions: we insert the set_grad at the beginning
+    grad_regions: OrderedDict[Node, Set[int]] = OrderedDict()
+
+    # For autocast regions:
```

Comment:
why is autocast different than grad? (why turn autocast off at the end of first region)

Reply:
Autocast behaves differently - it is context manager style i.e. exiting at the call to _exit_autocast - so we want to handle nested.

Grad is just mutating the global state. So it owns the entirety of the subsequent region until next set grad.

Reply:
Technically only 1 autocast needs to own an intermediate region, but to simplify the reasoning we allow multiple ownership.